src/modules/flow/iio/nodes.c

```diff
@@ -645,22 +653,22 @@ color_reader_cb(void *data, struct sol_iio_device *device)
     struct sol_flow_node *node = data;
     struct color_data *mdata = sol_flow_node_get_private_data(node);
     struct sol_rgb out = {
-        .red_max = mdata->out_range.max,
-        .green_max = mdata->out_range.max,
-        .blue_max = mdata->out_range.max
+        .red_max = mdata->iio_config.out_range.max,
+        .green_max = mdata->iio_config.out_range.max,
+        .blue_max = mdata->iio_config.out_range.max
     };
```

Comment:
You can simplify use of `color_data`, `double_data` and `direction_vector_data` if you do all of them extending a `base_data` - you called this one of `iio_device_config`. What I mean is:

```c
struct base_data {
    struct sol_iio_config config;
    struct sol_drange_spec out_range; // Not sure if it can be shared with color node, but currently is
    struct sol_iio_device *device;
    bool buffer_enabled : 1;
    bool use_device_default_scale : 1;
    bool use_device_default_offset : 1;
}

struct double_data {
    struct base_data base;
    struct sol_iio_channel *channel_val;
    double scale;
    double offset;
}
  
struct direction_vector_data {
    struct base_data base;
    struct sol_direction_vector scale;
    struct sol_direction_vector offset;
    struct sol_iio_channel *channel_x;
    struct sol_iio_channel *channel_y;
    struct sol_iio_channel *channel_z;
}

struct colour_data {
    struct base_data base;
    struct sol_iio_channel *channel_red;
    struct sol_iio_channel *channel_green;
    struct sol_iio_channel *channel_blue;
    double red_scale; // Or you can group `scale` and `offset` on a struct (and maybe even `channel`)
    double red_offset;// like you did. If you go that way, remember to name such struct as I said on
    double green_scale; // another comment
    double green_offset;
    double blue_scale;
    double blue_offset;
 }
```
Then you can unify `close` methods:

```c
static void
common_close(struct sol_flow_node *node, void *data)
{
    struct base_data *mdata = data;

    free((char *)mdata->config.trigger_name);
    if (mdata->device)
        sol_iio_close(mdata->device);
}
```

And on methods where you need the 'extended' struct, you use like

```c
   struct double_data *mdata = data;
   mdata->base.buffer_enabled = opts->buffer_size > -1;
```

Actually, it'll simplify uniting `tick` methods too.
What do you think?